File: Contests/CodeforcesPython/Educational Codeforces Round 132 Rated for Div. 2/B_Also_Try_Minecraft/B_Also_Try_Minecraft.py
n, m = map(int, input().split())
a = list(map(int, input().split()))

r = [[0] * n for _ in range(2)]
k1 = 0
k2 = 0
for i in range(1, n):
    k1 += max(a[i-1] - a[i], 0)
    r[0][i] = k1
    ni = n - i - 1
    k2 += max(a[ni+1] - a[ni], 0)
    r[1][ni] = k2
    
for i in range(m):
    s, t = list(map(int, input().split()))
    # print(s, t)
    # step = 1 if s < t else -1
    # d = 0
    # for j in range(s-1, t-1, step):
    #     # print(j)
    #     d += max(a[j] - a[j+step], 0)
    index = 0 if s < t else 1
    print(r[index][t-1] - r[index][s-1])

File: Contests/CodeforcesPython/Educational Codeforces Round 132 Rated for Div. 2/B_Also_Try_Minecraft/B_Also_Try_Minecraft.java
import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.*;
import java.util.function.IntUnaryOperator;
import java.util.function.LongUnaryOperator;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class B_Also_Try_Minecraft {
    public static void main(String[] args) {
        Scanner scn = new Scanner(System.in);
        int n = scn.nextInt();
        int m = scn.nextInt();
        int [] arr = new int [n];
        for(int i=0; i < n; i++) {
            arr[i] = scn.nextInt();
        }
        for(int i=0; i < m; i++) {
            int s = scn.nextInt();
            int t = scn.nextInt();
            int step = s < t ? 1:-1;
            int d = 0;
            int j = s-1;
            while(j!=t-1) {
                d += Math.max(arr[j] - arr[j+step], 0);
                j += step;
            }
            System.out.println(d);
        }
    }
}
